### PR TITLE
Rename Event resource as CTA

### PR DESCRIPTION
**Pull request by bnd5k**

### CODE REVIEW

File: config/database.yml [hunk 1]
```diff
@@ -12,7 +12,6 @@ development:
 # Do not set this db to the same as development or production.
 test: &test
   <<: *default
-  database: db/test.sqlite3
```
Comment (treznick) [inline config/database.yml]: whoops! maybe we should pull this out as its own commit?

File: app/models/cta.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  CTA_TYPES = { onsite: 1, phone: 2 }
+
+  belongs_to :contact, optional: true
+  belongs_to :location, optional: true
+  belongs_to :call_script, optional: true
+
+  validates_presence_of :title, :description, :website, :start_at, :cta_type
+  validate :validate_unique_cta, on: :create
+
+  scope :upcoming, -> { where("start_at >= ?", Date.today).order("start_at") }
```
Comment (treznick) [inline app/models/cta.rb]: I'd use `DateTime.current` instead of `Date.today` here to avoid using the server's timezone instead of the application's timezone (which we should set to `utc` in `application.rb`).

Also we should consider indexing the `start_at` column to avoid a full table scan with this scope.

File: app/models/cta.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  attr_accessor :cta_type
```
Comment (treznick) [inline app/models/cta.rb]: Do we still need the `attr_accessor` since you're overriding the getter and setter below?

File: app/resources/v1/cta_resource.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    filter :cta_type, apply: ->(records, value, _options) {
+      records.where(action_type: Event::CTA_TYPES[value[0].to_sym])
+    }
+
+    def start_time
+      @model.start_at.to_i
+    end
+
+    def start_time=(start_at)
+      @model.start_at = DateTime.strptime(start_at, "%s")
```
Comment (treznick) [inline app/resources/v1/cta_resource.rb]: are we expecting to receive datetime inputs in number of seconds? My assumption would that we'd be receiving ISO8601 strings? Ditto for the getters. Do we have any specifications as to what the standards for time formatting should be? Honest question.

File: db/migrate/20170331041221_create_ctas.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+      t.string :website
+      t.boolean :free, default: true
+      t.datetime :start_at
+      t.datetime :end_at
+      t.integer :action_type
+      t.uuid    :location_id
+      t.uuid    :contact_id
+      t.uuid    :call_script_id
+
+      t.timestamps
```
Comment (treznick) [inline db/migrate/20170331041221_create_ctas.rb]: We should add an index on the `start_at` column and a compound index on the columns used for uniqueness validation.

Comment (bnd5k) [inline db/migrate/20170331041221_create_ctas.rb]: Good call.

I've added that compound index.  I had to revise the criteria for uniqueness on the CTA model b/c Rails limits the length of the index name to 64 characters.   There are fewer fields we evaluate, but I think they're sufficient to exclude any potential duplicates.

Comment (treznick) [inline db/migrate/20170331041221_create_ctas.rb]: nice! Yeah, that 64 character index limit is a rough one. The only way I've seen to push that is to recompile postgres, which is about as close to the bottom of my "What I want to do as a web developer" list as is possible.

FWIW though you can pass the `name: <value>` keword argument to the `add_index` method in a migration to specify your own index name.

At one point in a past job I tried indexing the `reg_advisor_alert_specific_compliance_responsibilites_assignment_id` column. You probably get where that story is going...

Comment (bnd5k) [inline db/migrate/20170331041221_create_ctas.rb]: I think the criteria we've got in this PR are sufficient for determining whether a CTA is unique.  Let me know if you disagree.

Comment (treznick) [inline db/migrate/20170331041221_create_ctas.rb]: 👍
